### PR TITLE
Dto enhancement

### PR DESCRIPTION
**Pull request by supragyan**
Did code change in DTO Enhancement.
1.Did change start index and last index values in JSONService.svc.cs page.
2.Did code refactor in DTTable class.

### CODE REVIEW

File: root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                    { name: 'QuantityPerUnit', sortable: false, width: 150 },
+                    { name: 'UnitPrice', sortable: false, width: 50 },
+                    { name: 'UnitsInStock', sortable: false, width: 150 },
+                    { name: 'UnitsOnOrder', sortable: false, width: 150 },
+                    { name: 'ReorderLevel', sortable: false, width: 100 },
+                    { name: 'Discontinued', sortable: false, width: 150 }
+
+                ],
+                pager: $('#pager'),
+                rowNum: 29,
```
Comment (yusukemaegawa) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: 29 is not a round number.
Please specify 30.

File: root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                sortname: 'ProductID',
+                viewrecords: true,
+                sortorder: 'desc',
+                loadonce: false,
+                beforeProcessing: function (data, status, xhr) {  // レスポンスの受信時に呼び出されるイベント
+                    // サーバからエラーが返ってきた場合にダイアログを表示し、以降の処理を中止する
+                    // エラー時には、サーバからuserdataに情報を格納する
+                    if (data.userdata == "error") {
```
Comment (yusukemaegawa) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: "data.userdata" is undefined.
Therefore, this error handling is not correct.
(There is a error in original sample.)

Please compare "parseInt(data.page)" and "data.total".

"data.page" is requested page number.
"data.total" is total number of pages.

Comment (supragyan) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: >"data.userdata" is undefined.
 >Therefore, this error handling is not correct.
> (There is a error in original sample.)
Dear Maegawa-San,
HODC did check in ProjectX_Sample but there is no error.
But HODC did not check in Original Sample.
Do we need to check in Original Sample or do we need to implement for
comparing "parseInt(data.page)" and "data.total". in testJQGridJSON page?
Please confirm.

Comment (supragyan) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: Dear Maegawa-San,
\>\>If you request 5th page when total number of pages is 3, does the error dialog open?
No error dialogue box not showing but in grid.locale-da.js file it handled this issue i.e if empty records found then it showing some message.
Do we need to handle this  through error dialog?
Please confirm.

Comment (supragyan) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: Dear Maegawa-San,
\>\>If you request 5th page when total number of pages is 3, does the error dialog open?
\>\>I'd like you to revise to compare "parseInt(data.page)" and "data.total" in testJQGridJSON.aspx.
But HODC thought parseint is not required.
In testJQGridJSON.aspx. page,
if ((data.page) > (data.total)) {
                        alert("最大ページ以上を指定しています");
                        return false;
                    }
without parseint the above code is also working fine.
Please confirm.

Comment (supragyan) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: Please confirm.

Comment (yusukemaegawa) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: Sorry.
It seems that JavaScript converts string value to integer value implicitly when JavaScript compares string value and integer value. (data.page is string value.)

Therefore, "parseInt" is unnecessary.

Comment (yusukemaegawa) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: I'd like you to notify user if user inputs an invalid value.

As you say, the message is shown by "emptyrecords" property if user requests 5th page when the total number of pages is 3.
But, this message is also shown when the total number of rows in the table of database is 0.

Therefore, I'd like you to detect an invalid value and notify user by error dialog.

Comment (supragyan) [inline root/programs/C#/Samples/WebApp_sample/ProjectX_sample/Aspx/testFxLayerP/table/testJQGridJson.aspx]: Dear Nishino-San,
HODC did code commit for above issue.
HODC requesting to review and confirm.

If we need to create pull request then please close '
https://github.com/SymphonyTeleca/OpenTouryo/pull/34'

Please confirm.

**Comment by daisukenishino** [comment]
Maekawa-san said that it has been completed.
I want to merge this pull request.
However, there is a possibility that the pull request is old.
We request you to change the feature branch to newest
 by pulling nishino branch. and, request you to resend the pull request.

**Comment by supragyan** [comment]
Sure Nishino-San.

**Comment by supragyan** [comment]
Dear Nishino-San,

We pulled latest from nishino branch and did code commit to DTO Enhancement branch. 
Please close existing pull request 'https://github.com/SymphonyTeleca/OpenTouryo/pull/34'
So that we can create pull request again.

Regards,

**Comment by supragyan** [comment]
Dear Nishino-San,
Thank you for merging the pull request.
\>\>However, there is a possibility that the pull request is old.
\>\>We request you to change the feature branch to newest
\>\>by pulling nishino branch. and, request you to resend the pull request.
But this merged pull request is previous pull request.
Is this ok or do we need to create new pull request.

Regards,

**Comment by daisukenishino** [comment]
I missed the operation.
However, see below, it does not seem to have been rolled back in the old code.

https://github.com/SymphonyTeleca/OpenTouryo/blob/DTOEnhancement/root/programs/C%23/Samples/WebApp_sample/ProjectX_sample/Framework/Js/ie_key_event.js

https://github.com/SymphonyTeleca/OpenTouryo/blob/nishino/root/programs/C%23/Samples/WebApp_sample/ProjectX_sample/Framework/Js/ie_key_event.js

https://github.com/OpenTouryoProject/OpenTouryo/blob/develop/root/programs/C%23/Samples/WebApp_sample/ProjectX_sample/Framework/Js/ie_key_event.js

I took the changes of the # 68 as an example.
Enhancement of general-purpose DTO. #68

**Comment by supragyan** [comment]
Dear Nishino-San,
Sorry for inconvenience.
We did commit and created pull request as https://github.com/SymphonyTeleca/OpenTouryo/pull/49
Please merge and let us know.

Regards,

**Pull request merged**